### ISSUE TITLE
polyfill throws error in IE 11 when used with webcomponent transpiled by babel 7

### ISSUE THREAD

**Issue by franktopel** [WebReflection/document-register-element#145]
```log
5007: Die Eigenschaft "create" eines undefinierten oder Nullverweises kann nicht abgerufen werden.
document-register-element.js (2,2855)
```

![image](https://user-images.githubusercontent.com/10668603/42512640-6f9c41bc-8455-11e8-93f6-769cf21527d3.png)

Demo project attached. Load `index.babel.html` in IE 11.

`.babelrc` is included.

The Javascript is transpiled using the following steps:

1. `npm i -g @babel/cli`
2. `babel js/index.js --out-file js/index.transpiled.js`

### Expected behaviour: 

Both web components `<div is="my-div"></div>` and `<my-custom-div></my-custom-div>` render in IE 11 with no errors on the console. You can see they render if both are visible and have text content starting with `I am a beautiful...`

### Current behaviour:

Both web components `<div is="my-div"></div>` and `<my-custom-div></my-custom-div>` render in IE 11. Console shows the error mentioned above (sorry for German error message text, it says "cannot access property 'create' of null or undefined".

[wc-test.zip](https://github.com/WebReflection/document-register-element/files/2180744/wc-test.zip)

**Comment by WebReflection** [comment]
Hi. Babel is broken. You need Babel 7 or this to make it work:
https://github.com/WebReflection/babel-plugin-transform-builtin-classes

**Comment by franktopel** [comment]
As mentioned in the title, I am using babel 7.

**Comment by WebReflection** [comment]
your transpiled script doesn't work even with Chrome and native support for custom elements, with or without this polyfill.

When you'll have a working JS in Chrome that won't work with this polyfill I can have a look, until that day, you are using Babel beta and probably you should file a bug in there, removing this polyfill from the equation.

**Comment by WebReflection** [comment]
TL;DR this doesn't work, my poly isn't even there:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Web component test</title>
    <script src="js/index.transpiled.js"></script>
</head>
<body>
    <div is="my-div"></div>
    <my-custom-div></my-custom-div>
</body>
</html>
```

**Comment by franktopel** [comment]
Nonetheless it's your poly that throws these errors. Is that not worth looking into?

**Comment by franktopel** [comment]
I have just re-tested the file, and in my Chrome the web components both work, with and without your poly being referenced. Please re-check: 
[wc-test.zip](https://github.com/WebReflection/document-register-element/files/2181081/wc-test.zip)

**Comment by franktopel** [comment]
This version has working web components in IE 11 now. The error on the console remains, though.

**Comment by franktopel** [comment]
Also, why would I raise an issue (and this definitely is an issue) with Babel 7 Beta, if it's your polyfill that throws the error?

**Comment by WebReflection** [comment]
> webcomponents no longer working in Chrome 67

I use web components daily and with this polyfill. Maybe you should stop targeting ES5.

https://itnext.io/a-bloatless-web-d4f811c7991b

**Comment by franktopel** [comment]
As you might imagine, that decision is not up to me to make in that project. IE 11 is the current target platform, and will probably remain just that for as long as Microsoft does not declare end-of-life for that piece of software (which cannot happen before 2023 due to MS lifecycle policy).

**Comment by WebReflection** [comment]
Sure, but that shouldn't penalize every other evergreen browser. If you read my article it explains how to split bundles for any IE vs every other browser.

You can also consider the Universal Bundle Loader approach, described in here:
https://medium.com/@WebReflection/a-universal-bundle-loader-6d7f3e628f93

**Comment by franktopel** [comment]
The company allowed me **one day** to fix the issue, otherwise the decision is to no longer support anything other than IE 11. In an ideal world alot of things would be done differently. Unfortunately the real world could hardly be further away from being ideal in most scenarios/projects/cases.

**Comment by WebReflection** [comment]
If I understand correctly everything works, and you just have an error in console.

I will try to fix this IE11 issue but I cannot guarantee when.

**Comment by franktopel** [comment]
It works in the mini demo repo, it does not work in our project. But I'm hoping that is related to the errors in the console.

**Comment by WebReflection** [comment]
<img width="818" alt="screen shot 2018-07-11 at 15 31 28" src="https://user-images.githubusercontent.com/85749/42574802-99d8f1a8-851f-11e8-98dd-961eb73ac368.png">

everything works as expected. The only change I've made is to your code, because you write constructors even when you don't need it.

empty constructors are useless in JS, but if you need the constructor for whatever reason, you need to do it right, returning whatever the `super` returned, and also passing arguments.

This is well explained in here:
https://github.com/WebReflection/document-register-element#v1-caveat

Your code can be like this:
```js
// customized built-in
class MyDiv extends HTMLDivElement {
  connectedCallback() {
    console.log('customized built-in here');
    this.innerHTML = "I am a beautiful customized built-in element webcomponent!";
    this.addEventListener('click', () => {
      console.log(`clicked ${this.tagName}`);
    });
  }
}

// autonomous custom element
class MyCustomDiv extends HTMLElement {
  connectedCallback() {
    console.log('autonomous custom element here');
    this.innerHTML = "<div>I am a beautiful autonomous custom element webcomponent!</div>";
    this.addEventListener('click', () => {
      console.log(`clicked ${this.tagName}`);
    });
  }
}
```

otherwise, you need to do this:
```js
class MyDiv extends HTMLDivElement {
  constructor(...args) {
    const self = super(...args);
    // do whatever you need to do ... and *then*
    return self;
  }
  // ...
}
```

this is a documented caveat of the polyfill but also how JS classes / inheritance work in general, because a `super` you don't own, could always return something else that is not necessarily the same context.

In this case the polyfill needs to update the DOM node, which is returned by the super.

Regards.

**Comment by franktopel** [comment]
I guess one shouldn't trust documentation on the web too much.

> https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements

```javascript
class WordCount extends HTMLParagraphElement {
  constructor() {
    // Always call super first in constructor
    super();

    // Element functionality written in here

    ...
  }
}
```

**Comment by WebReflection** [comment]
> Maybe you want to to correct MDN?

if elements are native, and the syntax is native, you don't need to do the super dance the way I've explained it.

There are three different issues here:

  * a constructor that contains only a super call is useless
  * if the code is not transpiled and there are no polyfills with documented caveats, that syntax works, even if not perfect
  * by default, the internal operation performed by JS classes, if a constructor is not provided/used is the following one: `constructor(...args) { return super(...args); }`

I don't have time to maintain MDN documentation, I fix obvious mistakes when I find them, but this is not straight forward to explain.

However, I have written a little documentation in here too, so maybe next time read also the project README, not just MDN 😉

**Comment by franktopel** [comment]
> if the code is not transpiled and there are no polyfills with documented caveats

I don't know which universe that would work in. I've never met a real world project where this would be applicable.

**Comment by WebReflection** [comment]
FWIW I've added a note in the documentation.

**Comment by franktopel** [comment]
Interestingly there's no errors at all in Chromium 67 on Arch Linux, even without any `constructor()` changes. Can this be related to the OS?

**Comment by franktopel** [comment]
Even when I adjust teh constructor like you suggested above:

```javascript
    constructor(...args) {
        const self = super(...args);
        self.tableheight = null;
        self.tabletitle = null;
        self.dataSelectionMode = null;
        self.dataServersidePagination = 'false';
        return self;
    }
``` 
I'm still getting an error in the console (and the browser even hangs):

![image](https://user-images.githubusercontent.com/10668603/42622141-fff2e12e-85bf-11e8-9859-5abe29efafae.png)

Would you, by any chance, be available short-term for hiring for a week or two?

**Comment by WebReflection** [comment]
> Would you, by any chance, be available short-term for hiring for a week or two?

I'm afraid I'm a full time employee already so the answer is no.

In any case, I have no idea how you can blame this polyfill at line 44592 and column 222 ... try to debug your code and provide a better message or I won't ever be able to help, not even hired.

**Comment by franktopel** [comment]
This is the line of code I'm directed to when I click on `toolkit.js`:

```javascript
function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null) return null; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }
```

**Comment by WebReflection** [comment]
@franktopel I am not going to debug this for you, I have quite a busy day ahead to be honest.

> May I ask who is your employer?

I don't feel comfortable in you contacting them for anything related to this polyfill but my employer is no secret: https://eyeo.com

**Comment by franktopel** [comment]
Maybe we can get in touch via email? info@connexo.de is my email address.
